### PR TITLE
[babel 8] Reland "Use `NodePath#hub` as part of the paths cache key"

### LINKED ISSUE #6437: Figure out an approach to NodePath 'hub' object tracking with better cache behavior
I recently realized we've got bug with our invalidation logic for NodePath that can cause some real weirdness. Here it is [on ASTExplorer](http://astexplorer.net/#/gist/232f9956e84a89d90fd02b013071778a/797e7d58601ed6327610784494f19bb18197a9f5) and here:

```
var mem = template("property.access")();

traverse(mem, {noScope: true});

var p2 = path.insertAfter(mem)[0];

console.log(p2.get("expression").hub);
```
will print `undefined` for `hub` even through `p2.hub` is the `Hub` class instance. 

Which is surprising because in https://github.com/babel/babel/blob/d511cfc0e287d3d33292d7f7afa035c37e51b980/packages/babel-traverse/src/path/index.js#L74, the `hub` value is explicitly supposed to pass down.

It fails in this case because on https://github.com/babel/babel/blob/d511cfc0e287d3d33292d7f7afa035c37e51b980/packages/babel-traverse/src/path/index.js#L91 if a value is found in the cache, we never bother bailing out.

The problem is, what do we do. Two options come to mind:
1. Overwrite the `.hub` on the cache result.
2. Only accept items from the cache if the hub is correct.

#### Option 1

Pretty  gross because we could just as easily be introducing another bug by removing a Hub instance that was supposed to be there

#### Option 2
Sounds reasonable, until you realize that we actually _crucially_ rely on this gross caching behavior to instantiate `NodePath` with the correct `Hub` instance in the first place.

In `File` init in https://github.com/babel/babel/blob/d511cfc0e287d3d33292d7f7afa035c37e51b980/packages/babel-core/src/transformation/file/file.js#L42 we create the root `NodePath`, but then we never really do anything with it. Traversal begins on https://github.com/babel/babel/blob/d511cfc0e287d3d33292d7f7afa035c37e51b980/packages/babel-core/src/transformation/index.js#L71, which never explicitly mentions `this.path` at all. That path ends up being used, thus using the root `Hub`, because it is present in the cache. If we explicitly test for the `Hub` in the cache, we'll break this weird cache-based pass-through logic.

#### Solution?

To me, the "feature" we're breaking in option 2 is totally a mistake to rely on since it is essentially relying on sideeffects to the extreme. This means we need a way to do two things:

1. Initialize the NodePath up front, so it can be stored on the `File` class instance
2. Perform a traversal using this NodePath, while ensuring that same traversal logic we have now is preserved.

The exact details on how we do those two things is up in the air, but feels like something we should figure out ASAP.

Incidentally, this same bug is the reason we have to do https://github.com/babel/babel/blob/d511cfc0e287d3d33292d7f7afa035c37e51b980/packages/babel-template/src/index.js#L218 to explicitly clear the cache, otherwise every node we get pack from `babel-template` would trigger this bug, due to the internal traversal, which doesn't have a `hub`, polluting the cache.

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #6437
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This is just a revert of the revert of #15725, it's not fixed yet.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/15759"></a>

### CODE REVIEW

**Comment by nicolo-ribaudo** [comment]
The angular error is caused by the caching changes, and not by the changes in `babel-core` to traverse the `Program` node with the existing parent path.

**Comment by nicolo-ribaudo** [comment]
Ohh :/

The problem is caused by using an old version of `@babel/core` with newer versions of `@babel/traverse`, since older versions of Babel core obviously do not have the changes in [packages/babel-core/src/transformation/index.ts](https://github.com/babel/babel/pull/15759/files#diff-72f02629256ac540c80675cb14725428acc768cd198333f31ab086702066430e).

**Comment by nicolo-ribaudo** [comment]
This should fix the compat problem, tested locally. Let's see if it fixes the angular-cli integration test.

The fix/check is _very bad_, but it's self-contained and designed to run as less as possible (it only runs when calling into `@babel/traverse` through one of the methods used by `@babel/core`, and only when those methods are not called recursively via `@babel/traverse`).

File: packages/babel-traverse/src/path/index.ts [hunk 1]
```diff
@@ -92,12 +92,57 @@ class NodePath<T extends t.Node = t.Node> {
...
+          // noHubInCacheKeyForBackwardCompat has three states (true, false, and
+          // undefined) instead of two so that we can run the magic checks only
+          // at the top-level of the traverse call stack, and not when recoursing.
+          // new Error().stack is expensive, and we want to avoid it as much as
+          // possible in this very hot path.
+          // See the .noHubInCacheKeyForBackwardCompat definition in ./cache.ts
+          // for why we need this.
+          // @ts-expect-error ts does not know about this property
+          if (cache.noHubInCacheKeyForBackwardCompat === undefined) {
+            const callerIsOldBabelCore =
```
Comment (nicolo-ribaudo) [inline packages/babel-traverse/src/path/index.ts:107]: Uhm, this returns `true` even for new babel versions, I have to find an alternative.

**Comment by alan-agius4** [comment]
In general we pin all of our direct dependencies to have a more stable releases. By pinning all of our dependencies we reduce compatibility issues, bugs and other breaky behaviour as some packages don’t follow semver. Over the years this did prove to make the Angular CLI more stable.

However, in this case though pinning did cause a problem.

**Comment by nicolo-ribaudo** [comment]
This is ready for review, but the code is _incredibly_ ugly.

Performance wise it's ok: the `new Error().stack` check to see if our caller is `@babel/core` only happens once (per compiled file) when calling `@babel/traverse` from old versions of `@babel/core` or from programs using `@babel/traverse` directly.

The angular-cli (breaking) test is failing because the workaround only applies to Babel 7, but it will start passing once `@angular/cli` updates their `@babel/core` version.

**Review by JLHwung — APPROVED** [review]
🤦

**Comment by liuxingbaoyu** [comment]
Could we possibly make it only land in babel8?
The method based on stack detection doesn't seem very elegant, especially if someone disabled the stack.
This bug has been around for ages, and it should be acceptable to defer to babel8. Even I worry that someone is relying on the wrong behavior. 🤦‍♂️

**Comment by nicolo-ribaudo** [comment]
I updated the PR to only land the breaking change in Babel 8.

**Review by liuxingbaoyu — APPROVED** [review]
Thanks!

**Pull request merged**